src/libsync/propagatedownload.h

```diff
@@ -118,6 +124,35 @@ private slots:
     void slotMetaDataChanged();
 };
 
+/**
+ * @brief The GETEncryptedFileJob class that provides file decryption on the fly while the download is running
+ * @ingroup libsync
+ */
+class OWNCLOUDSYNC_EXPORT GETEncryptedFileJob : public GETFileJob
+{
+    Q_OBJECT
+
+public:
+    // DOES NOT take ownership of the device.
+    explicit GETEncryptedFileJob(AccountPtr account, const QString &path, QIODevice *device,
+        const QMap<QByteArray, QByteArray> &headers, const QByteArray &expectedEtagForResume,
+        qint64 resumeStart, EncryptedFile encryptedInfo, QObject *parent = nullptr);
+    explicit GETEncryptedFileJob(AccountPtr account, const QUrl &url, QIODevice *device,
```

Comment:
not sure if it is possible but you could create static factory methods that describe for which cases the different constructors can be used. At the moment it's very difficult to say what the difference between them is.

Reply:
@FlexW These are copy-pasted from `GETFileJob`, so, that one needs refactoring too if we are going to use patterns. Why is it difficult to see the difference? One is accepting `QUrl`, while another one is accepting `QString`.

Reply:
@allexzander Can we not start incrementally making changes? I mean that code is new. Why not write it in a clearer way?

Reply:
@FlexW Could do, but, how would those factory methods be called then?

`static GETEncryptedFileJob *createEncryptedFileJobFromStringPath()`
and
`static GETEncryptedFileJob *createEncryptedFileJobFromURL()`
?

If I got you right, that's the difference that we must clearly state, when adding factory methods?

Then, how do we decide which one to use and when? The one that accepts a `QUrl `and the one accepting `QString`?

An example would be appreciated if possible.

Reply:
> If I got you right, that's the difference that we must clearly state, when adding factory methods?

Yes :)

Just a minor thing I would name it `FromUrl` instead of `FromURL`. 

> Then, how do we decide which one to use and when? The one that accepts a QUrl and the one accepting QString?

I think you created these two different constructors for a reason? :) So you should know where you need to use each? But I think you would use the QString at anytime you just have a string and the QUrl at anytime where you just have a QUrl.